RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+
+  for(size_t ihit = 0; ihit<ecalrechitcoll->size(); ++ ihit){
+    HaloClusterCandidateEE  clustercand;
+    
+    const EcalRecHit & rechit = (*ecalrechitcoll)[ ihit ];
     math::XYZPoint rhpos = getPosition(rechit.id(),vtx);
+    //Et condition
     double rhet = rechit.energy()/cosh(rhpos.eta());
-    double dphi_rhseg = abs(deltaPhi(rhpos.phi(),iPhi));
-    double dr_rhseg = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y()) - iR;
-    double dtcorr_rhseg = rechit.time()- abs(rhpos.z()-iZ)/30- iT; 
-    if( 
-       ( rechit.time()<-1)&&
-       (rhpos.z()*iZ<0|| abs(rhpos.z())<200)&&
-       rhet> et_thresh_rh&&
-       dphi_rhseg < dphi_thresh_segvsrh &&
-       dr_rhseg < dr_highthresh_segvsrh && dr_rhseg> dr_lowthresh_segvsrh && //careful: asymmetric cut might not be the most appropriate thing 
-       dtcorr_rhseg> dt_lowthresh_segvsrh && dtcorr_rhseg< dr_highthresh_segvsrh
-       ) return true; 
+    if(rhet<et_thresh_seedrh) continue;
+    double eta = rhpos.eta();
+    double phi = rhpos.phi();
+    double rhr = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y());
+    
+    bool isiso = true;
+    double etcluster(0);
+    double timediscriminator(0);
+    int clustersize(0);
+    int nbcrystalssmallt(0);
+    int nbcrystalshight(0);
+    //Building the cluster
+    edm::RefVector<EcalRecHitCollection> bhrhcandidates;
+    for(size_t jhit = 0; jhit<ecalrechitcoll->size(); ++ jhit){
+      const EcalRecHit & rechitj = (*ecalrechitcoll)[ jhit ];
+      EcalRecHitRef rhRef(ecalrechitcoll,jhit);
+      math::XYZPoint rhposj = getPosition(rechitj.id(),vtx);
+
+      //Ask the hits to be in the same endcap
+      if(rhposj.z()*rhpos.z()<0)continue;
+
+      double etaj = rhposj.eta();
+      double phij = rhposj.phi();
+      double dr = sqrt(abs(eta-etaj)*abs(eta-etaj)+deltaPhi(phi,phij)*deltaPhi(phi,phij));
+
+      //Outer cone
+      if(dr>0.3) continue;
+      
+      double rhetj = rechitj.energy()/cosh(rhposj.eta());
+      //Rechits with et between 1 and 2 GeV are saved in the rh list but not used in the calculation of the halocluster variables
+      if(rhetj<1) continue; 
+      bhrhcandidates.push_back(rhRef);
+      if(rhetj<2) continue;
+      
+      //Isolation between outer and inner cone
+      if(dr>0.05){isiso=false;break;}//The deposit should be isolated
+
+      etcluster+=rhetj;
+  
+      //Timing infos:
+      //Here we target both IT and OT beam halo 
+      double rhtj=rechitj.time();
+
+      //Discriminating variables for OT beam halo: 
+      if(rhtj>1) nbcrystalshight++;
+      if(rhtj<0) nbcrystalssmallt++;
+      //Timing test (likelihood ratio), only for seeds with large R (100 cm) and for crystals with et>5, 
+      //This targets IT beam halo (t around - 1ns) 
+      if(rhtj>5){
+      double corrt_j = rhtj + sqrt(rhposj.x()*rhposj.x()+rhposj.y()*rhposj.y() + 320.*320.)/30. - 320./30.;
+      timediscriminator+=log( TMath::Gaus(corrt_j,0,0.4,false)/TMath::Gaus(corrt_j,0.3,0.4,false) );
```

Comment:
Please reduce to the 2nd order polynomial of corrt_j.
There is no need to do ``` log (exp ( - corrt_j*corrt_j/0.32) ) ``` just to loose precision and possibly generate NaNs on the way.